lib/usd/pxrUsdPreviewSurface/usdPreviewSurfaceReader.cpp

```diff
@@ -0,0 +1,160 @@
+//
+// Copyright 2018 Pixar
+// Copyright 2020 Autodesk
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+#include "usdPreviewSurface.h"
+
+#include <mayaUsd/fileio/shaderReader.h>
+#include <mayaUsd/fileio/shaderReaderRegistry.h>
+#include <mayaUsd/fileio/translators/translatorUtil.h>
+#include <mayaUsd/fileio/utils/readUtil.h>
+#include <mayaUsd/utils/util.h>
+
+#include <pxr/base/tf/diagnostic.h>
+#include <pxr/base/tf/staticTokens.h>
+#include <pxr/base/tf/token.h>
+#include <pxr/base/vt/value.h>
+#include <pxr/pxr.h>
+#include <pxr/usd/sdf/path.h>
+#include <pxr/usd/sdf/types.h>
+#include <pxr/usd/sdf/valueTypeName.h>
+#include <pxr/usd/usdShade/input.h>
+#include <pxr/usd/usdShade/output.h>
+#include <pxr/usd/usdShade/shader.h>
+#include <pxr/usd/usdShade/tokens.h>
+
+#include <maya/MFnDependencyNode.h>
+#include <maya/MObject.h>
+#include <maya/MPlug.h>
+#include <maya/MStatus.h>
+
+PXR_NAMESPACE_OPEN_SCOPE
+
+class PxrMayaUsdPreviewSurface_Reader : public UsdMayaShaderReader {
+public:
+    PxrMayaUsdPreviewSurface_Reader(const UsdMayaPrimReaderArgs&);
+
+    bool Read(UsdMayaPrimReaderContext* context) override;
+
+    TfToken GetMayaNameForUsdAttrName(const TfToken& usdAttrName) override;
+};
+
+PXRUSDMAYA_REGISTER_SHADER_READER(UsdPreviewSurface, PxrMayaUsdPreviewSurface_Reader)
+
+PxrMayaUsdPreviewSurface_Reader::PxrMayaUsdPreviewSurface_Reader(
+    const UsdMayaPrimReaderArgs& readArgs)
+    : UsdMayaShaderReader(readArgs)
+{
+}
+
+/* virtual */
+bool PxrMayaUsdPreviewSurface_Reader::Read(UsdMayaPrimReaderContext* context)
+{
+    const auto&    prim = _GetArgs().GetUsdPrim();
+    UsdShadeShader shaderSchema = UsdShadeShader(prim);
+    if (!shaderSchema) {
+        return false;
+    }
+
+    MStatus           status;
+    MFnDependencyNode depFn;
+    if (!(UsdMayaTranslatorUtil::CreateShaderNode(
+              MString(prim.GetName().GetText()),
```

Comment:
Maya will automatically append a numeric increments to make sure the name is unique, but that might not be the best solution. I suspect we will have proponents for both options, which might require exposing that as an import option instead of going one way or the other.

On the same subject, it would be nice to produce less materials when possible: if we have the same cup/plate/crayon asset file coming from 2 different paths, it would be nice to be able to produce a single Maya material that link both geometries. The Kitchen_set demo scene has a nice two-toned tiled floor with should, in a perfect world, result in only 2 lambert materials when importing using the "Display color" shading mode.

Do you think this would be of interest?

Reply:
I think it's less about trying to ensure that the node names are unique, and more important that the translation is as consistent as possible across multiple roundtrips. When there are potentially existing overs on the prims being exported, it would be easy for those to fall off if a name collision on import caused the node name to change, which then affects the prim path when re-exported.

And yeah, I think material deduplication would be an interesting feature to pursue if we can detect that the "source" materials in USD are all the same material (e.g. bindings from multiple gprims across references that all resolve to what is ultimately a single material).